docs/RFCS/cluster_upgrade_tool.md

```diff
@@ -0,0 +1,76 @@
+- Feature Name: Cluster Upgrade Tool
+- Status: draft
+- Start Date: 2016-09-15
+- Authors: Daniel Harrison
+- RFC PR: (PR # after acceptance of initial draft)
+- Cockroach Issue: (one or more # from the issue tracker)
+
+# Summary
+
+A series of hooks to perform any necessary bookkeeping before a CockroachDB
+version upgrade.
+
+# Motivation
+
+Backward compatibility is crucial to a database, but can introduce considerable
+complexity to a codebase. All else being equal, it's better to keep backward
+compatibility logic siloed from the significant complexity that is inherent in
+software as complex as CockroachDB.
+
+# Detailed design
+
+Some versions of CockroachDB will require that a "system upgrade hook" be run
+before the cluster is upgraded to that version. This is expected to be
+infrequent; not every release will bump the version, only ones that need a
+migration hook.
+
+A new kv entry `SystemVersion` (ClusterVersion?) is introduced in the system
+keys to keep track of which of these hooks was the last to run. It stores a
```

Comment:
Agreed. I was wary of overengineering, but since you're concerned about this too, I added something about to address it. (Plus I tweaked it a bit so you don't have to know the cockroach version your code is going into, which might not always be straightforward).